src/main/java/net/neoforged/neoforge/transfer/transaction/SnapshotParticipant.java

```diff
@@ -0,0 +1,116 @@
+package net.neoforged.neoforge.transfer.transaction;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+
+/**
+ * A base participant implementation that modifies itself during transactions,
+ * saving snapshots of its state in objects of type {@code T} in case it needs to revert to a previous state.
+ *
+ * <h3>How to use from subclasses</h3>
+ * <ul>
+ *     <li>Call {@link #updateSnapshots} right before the state of your subclass is modified in a transaction.</li>
+ *     <li>Override {@link #createSnapshot}: it is called when necessary to create an object representing the state of your subclass.</li>
+ *     <li>Override {@link #readSnapshot}: it is called when necessary to revert to a previous state of your subclass.</li>
+ *     <li>You may optionally override {@link #onFinalCommit}: it is called at the of a transaction that modified the state.
+ *     For example, it could contain a call to {@code markDirty()}.</li>
+ *     <li>(Advanced!) You may optionally override {@link #releaseSnapshot}: it is called once a snapshot object will not be used,
+ *     for example you may wish to pool expensive state objects.</li>
+ * </ul>
+ *
+ * <h3>More technical explanation</h3>
+ *
+ * <p>{@link #updateSnapshots} should be called before any modification.
+ * This will save the state of this participant using {@link #createSnapshot} if no state was already saved for that transaction.
+ * When the transaction is aborted and changes need to be rolled back, {@link #readSnapshot} will be called
+ * to signal that the current state should revert to that of the snapshot.
+ * The snapshot object is then {@linkplain #releaseSnapshot released}, and can be cached for subsequent use, or discarded.
+ *
+ * <p>When an outer transaction is committed, {@link #readSnapshot} will not be called so that the current state of this participant
+ * is retained. {@link #releaseSnapshot} will be called because the snapshot is not necessary anymore,
+ * and {@link #onFinalCommit} will be called after the transaction is closed.
+ *
+ * @param <T> The objects that this participant uses to save its state snapshots.
+ */
+public abstract class SnapshotParticipant<T> implements Transaction.CloseCallback, Transaction.OuterCloseCallback {
+    private final List<T> snapshots = new ArrayList<>();
```

Comment:
Given we never seem to fully release memory claimed into the snapshots (we just set the elements back to null), would this list be better off as a `Int2ObjMap<T>`? Which given unless I missed anything would mean we only bother adding to the map when we have something, and otherwise can allow the default not present being null to take place. While yes we probably wouldn't bother resizing the map to be smaller, the odds of it reaching larger sizes would also be decreased.